bd-artifact-upload/src/uploader.rs

```diff
@@ -0,0 +1,666 @@
+    let files = self
+      .file_system
+      .list_files(&REPORT_DIRECTORY)
+      .await
+      .unwrap_or_default();
+
+    for file in files {
+      if file == REPORT_INDEX_FILE.to_string_lossy() {
+        continue;
+      }
+
+      if !filenames.contains(&file) {
+        log::debug!("removing artifact {file} from disk, not in index");
+        if let Err(e) = self
+          .file_system
+          .delete_file(&REPORT_DIRECTORY.join(&file))
+          .await
+        {
+          log::warn!("failed to delete artifact {file:?}: {e}");
+        }
+      }
+    }
+  }
+
+  async fn handle_intent_negotiation_decision(&mut self, decision: IntentDecision) {
+    match decision {
+      IntentDecision::Drop => {
+        self.stats.dropped_intent.inc();
+        let entry = &self.index.pop_front().unwrap();
+
+        if let Err(e) = self
+          .file_system
+          .delete_file(&REPORT_DIRECTORY.join(&entry.name))
+          .await
+        {
+          log::warn!("failed to delete artifact {:?}: {}", entry.name, e);
+        }
+
+        // Even if we failed to delete the artifact still try to clean up the index. There's a good
+        // chance this will fail too but might make it less likely that we try to re-upload this
+        // file.
+
+        self.index.pop_front();
+        self.write_index().await;
+      },
+      IntentDecision::UploadImmediately => {
+        self.stats.accepted_intent.inc();
+        let entry = self.index.front_mut().unwrap();
+        // Mark the file as being ready for uploads and persist this to the index.
+        entry.pending_intent_negotiation = false;
+        self.write_index().await;
+      },
+    }
+  }
+
+  async fn handle_upload_complete(&mut self) -> String {
+    self.stats.uploaded.inc();
+
+    let entry = self.index.pop_front().unwrap();
+    let file_path = REPORT_DIRECTORY.join(&entry.name);
+
+    if let Err(e) = self.file_system.delete_file(&file_path).await {
+      log::warn!("failed to delete artifact {:?}: {}", entry.name, e);
+    }
+
+    self.write_index().await;
+
+    entry.name
+  }
+
+  fn stop_current_upload(&mut self) {
+    if let Some(task) = self.upload_task_handle.take() {
+      task.abort();
+    }
+    if let Some(task) = self.intent_task_handle.take() {
+      task.abort();
+    }
+  }
+
+  async fn track_new_upload(&mut self, uuid: Uuid, contents: Vec<u8>, state: LogFields) {
+    // If we've reached our limit of entries, stop the entry currently being uploaded (the oldest
+    // one) to make space for the newer one.
+    if self.index.len() == usize::try_from(*self.max_entries.read()).unwrap_or_default() {
+      log::debug!("upload queue is full, dropping current upload");
+
+      self.stats.dropped.inc();
+      self.stop_current_upload();
+      self.index.pop_front();
+    }
+
+    let uuid = uuid.to_string();
+
+    // Add a CRC trailer to ensure we don't try to upload a malformed report.
+    let contents = write_checksummed_data(&contents);
```

Comment:
Per your other point about passing a file around, we could avoid another full copy of this memory here since we could stream it all from the file.